src/storage/src/source/types.rs

```diff
@@ -164,36 +159,27 @@ pub struct SourceOutput<K, V> {
     pub key: K,
     /// The record's value
     pub value: V,
-    /// The position in the partition described by the `partition` in the source
-    /// (e.g., Kafka offset, file line number, monotonic increasing
-    /// number, etc.)
-    pub position: MzOffset,
-    /// The time the record was created in the upstream system, as milliseconds since the epoch
-    pub upstream_time_millis: Option<i64>,
-    /// The partition of this message, present iff the partition comes from Kafka
-    pub partition: PartitionId,
-    /// Headers, if the source is configured to pass them along. If it is, but there are none, it
-    /// passes `Some([])`
-    pub headers: Option<Vec<(String, Option<Vec<u8>>)>>,
+    /// Additional metadata columns requested by the user
+    pub metadata: Row,
+    /// The offset position in the partition of a kafka source. This is field is on its way out and
+    /// its only valid use is in the upsert operator. Do NOT use it in any new place!
+    // TODO(petrosagg): remove this field
```

Comment:
what is the plan to remove the field?

Reply:
One of:
* Moving upsert before upsert, so that we can simply sort data based on the original timestamp
* Replacing it with something equivalent to `source_timestamp: Box<dyn Timestamp>` field that carries original timestamp before reclocking and allows sorting messages within a single timestamp after reclocking. We'll need to figure out a way that doesn't require allocation, maybe we can make it work with generics

I'm leaning towards the first one since it also has side benefits like reduced latency